enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,492 @@
+    - ##### NOTE: This step is optional for users directly creating CRs. They might already know what CRs they need for step 2.
+- PV Creation: After the user has decided how to created PVs, they can choose to create PVs with either the LocalVolume CR or the AutoDetectVolume CR.
+
+
+## Design Details for `Auto discovery of local devices`
+
+API scheme for `LocalVolumeDiscovery` and `LocalVolumeNodeDiscoveryResult` CR:
+
+```go
+type LocalVolumeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscovery `json:"items"`
+}
+
+type LocalVolumeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoverySpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoverySpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type LocalVolumeDiscoveryPhase string
+
+const (
+	InProgressPhase LocalVolumeDiscoveryPhase = "InProgress"
+	FailedPhase     LocalVolumeDiscoveryPhase = "Failed"
+	CompletedPhase  LocalVolumeDiscoveryPhase = "Completed"
+)
+
+type LocalVolumeDiscoveryStatus struct {
+	Phase         LocalVolumeDiscoveryPhase `json:"phase,omitempty"`
+	CompletedJobs *int32                    `json:"completedjobs,omitempty"`
+	TotalJobs     *int32                    `json:"totaljobs,omitempty"`
+}
+
+type LocalVolumeNodeDiscoveryResultList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeNodeDiscoveryResult `json:"items"`
+}
+
+type LocalVolumeNodeDiscoveryResult struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeNodeDiscoveryResultSpec   `json:"spec"`
+	Status            LocalVolumeNodeDiscoveryResultStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeNodeDiscoveryResultSpec struct {
+	NodeName string `json:"nodeName"`
+}
+
+type LocalVolumeDiscoveryResultPhase string
+
+const (
+	InProgressPhase LocalVolumeDiscoveryResultPhase = "Discovering"
+	FailedPhase     LocalVolumeDiscoveryResultPhase = "Failed"
+	CompletedPhase  LocalVolumeDiscoveryResultPhase = "Discovered"
+)
+
+type LocalVolumeNodeDiscoveryResultStatus struct {
+	Phase     LocalVolumeDiscoveryResultPhase `json:"phase,omitempty"`
+	TimeStamp string                          `json:"timestamp,omitempty"`
+	// Devices contains the list of devices on which LSO
+	// is capable of creating LocalPVs
+	// The devices in this list qualify these following conditions.
```

Comment:
If the discovery is re-run after the local PVs, will the same devices get returned? In other words, can we filter out the devices that have local PVs created? None of these conditions listed would filter out the local PVs. I know in a previous comment there was concern about being able to detect this. Was an answer found? If not, this needs more investigation.

Reply:
One way would be to use - https://golang.org/pkg/path/filepath/#EvalSymlinks
We know that symlinks for detected devices are created in `/mnt/local-storage/<storageClassName>`. So we can evaluate if a device already has a symlink created in this directory.
If yes, we can skip the device. We already use it in the existing "Local Storage Operator"  [code base](https://github.com/openshift/local-storage-operator/blob/3f4a84cafc3ecbe59a83428650c2eed60bca423a/pkg/diskmaker/diskmaker.go#L269) to check for stable device ID for a device.

Just a thought. Haven't tested it though